lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+                             RunIdentifierTy runId, DoneCbTy cb) {
+  // If list of roots is empty, there is nothing to do. Give back the
+  // context so the caller can reuse it.
+  if (!root) {
+    cb(runId, ResultCode::Executed, std::move(context));
+    return;
+  }
+
+  // If the given run ID corresponds to a run already in progress, there is
+  // also nothing to do, but return an error. Give back the context so the
+  // caller can reuse it.
+  std::unique_lock<std::mutex> lock(executionStatesMutex_);
+  if (executionStates_.find(runId) != executionStates_.end()) {
+    cb(runId, ResultCode::Failed, std::move(context));
+    return;
+  }
+
+  // Create execution state tracker object for this run ID.
+  auto executionState =
+      std::make_shared<ExecutionState>(runId, root, std::move(cb));
+  executionStates_.insert(std::make_pair(runId, executionState));
+  lock.unlock();
+
+  // Execute all child nodes of root.
+
+  // Mark the child nodes as "inflight" (i.e. currently executing). This must be
+  // done here instead of inside executeDAGNode() so that a node can be
+  // executed while placeholders are being propagated for the next node without
+  // the callback for that node deleting the execution state.
+  executionState->incrementInflightNodes((root->children).size());
+
+  // Convert the Context to shared_ptr.
+  std::shared_ptr<Context> sharedContext = std::move(context);
+
+  for (auto const &node : root->children) {
+    // Propagate placeholders from the given starter Context into the input
+    // Context for the current node being processed.
+    propagatePlaceholdersForNode(runId, node, sharedContext);
+
+    // Execute the node.
+    executeDAGNode(runId, node);
+  }
+}
+
+void ThreadPoolExecutor::propagatePlaceholdersForNode(
+    RunIdentifierTy runId, const DAGNode *node, std::shared_ptr<Context> ctx) {
+  // Get the execution state for the run.
+  auto it = executionStates_.find(runId);
+
+  if (it == executionStates_.end()) {
+    assert("Execution state should exist!");
+  }
+
+  std::shared_ptr<ExecutionState> executionState = it->second;
+
+  // Get the map of Placeholder -> Tensor mappings within ctx.
+  const Context::PlaceholderMap &phMap = ctx->pairs();
+
+  // Get the symbol table for the node.
+  const RuntimeBundle &bundle = node->runtimeBundle;
+  const SymbolTable &symbolTable = bundle.getSymbolTable();
+
+  for (const auto &symbolPair : symbolTable) {
+    const auto &symbolName = symbolPair.first;
+    const auto &symbolInfo = symbolPair.second;
+
+    if (symbolInfo.input) {
+      // If the symbol is an input, look for a Placeholder in ctx with
+      // the same name and copy the corresponding Tensor into the input Context
+      // being prepared for the node.
+      for (const auto &phTensorPair : phMap) {
+        auto *placeholder = phTensorPair.first;
+        const auto *tensor = phTensorPair.second;
+
+        if (symbolName == placeholder->getName()) {
```

Comment:
Rather than this N*M comparison loop, why don't we add name->PH as state in the context and just do one lookup?

Reply:
Will do.